### PR TITLE
Rewrite in ES6+

### PR DESCRIPTION
**Pull request by steelbrain**
- async file searching instead of sync
- rewrite in es6+

### CODE REVIEW

File: lib/main.js [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+          const errorLine = error.line - 1
+          let range
+
+          // TODO: Remove workaround of jshint/jshint#2846
+          if (error.character === null) {
+            range = Helpers.rangeFromLineNumber(textEditor, errorLine)
+          } else {
+            let character = error.character
+            const maxCharacter = textEditor.getBuffer().lineLengthForRow(errorLine)
+            // TODO: Remove workaround of jshint/jshint#1457
```
Comment (Arcanemagus) [inline lib/main.js]: This is actually esprima, not jshint.

File: lib/main.js [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+import { CompositeDisposable } from 'atom'
+
+type Linter$Provider = Object
+
+module.exports = {
+  config: {
+    executablePath: {
+      type: 'string',
+      default: Path.join(__dirname, '..', 'node_modules', 'jshint', 'bin', 'jshint'),
+      description: 'Path of the `jshint` executable.'
```
Comment (Arcanemagus) [inline lib/main.js]: Should note that this is the _Node_ executable, not the system (wrapper script) executable.

Comment (steelbrain) [inline lib/main.js]: I wonder why it was set to that in the previous version (I just copy/pasted the configs)

File: lib/main.js [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+      default: false,
+      description: 'Disable linter when no `.jshintrc` is found in project.'
+    },
+    lintJSXFiles: {
+      title: 'Lint JSX Files',
+      type: 'boolean',
+      default: false
+    }
+  },
+  activate() {
```
Comment (Arcanemagus) [inline lib/main.js]: Not a huge deal, but it's nicer to have the extra separation of a blank line between major sections like this.

File: lib/main.js [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+      } else {
+        if (this.scopes.indexOf(scopeJSX) !== -1) {
+          this.scopes.splice(this.scopes.indexOf(scopeJSX), 1)
+        }
+      }
+    }))
+
+    const scopeEmbedded = 'source.js.embedded.html'
+    this.subscriptions.add(
+      atom.config.observe('linter-jshint.lintJSXFiles', lintInlineJavaScript => {
```
Comment (Arcanemagus) [inline lib/main.js]: `linter-jshint.lintJSXFiles` -> `linter-jshint.lintInlineJavaScript`.

File: lib/main.js (hunk 2 above)
Comment (steelbrain) [inline lib/main.js]: Just found out that it is actually the executable, ,we don't require it, we spawn it

File: lib/main.js (hunk 1 above)
Comment (Arcanemagus) [inline lib/main.js]: Also, this should be above the `maxCharacter` determination, as that is part of the workaround.

File: lib/main.js [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+            const maxCharacter = textEditor.getBuffer().lineLengthForRow(errorLine)
+            // TODO: Remove workaround of jshint/jshint#1457
+            if (character > maxCharacter) {
+              character = maxCharacter
+            }
+            range = Helpers.rangeFromLineNumber(textEditor, errorLine, character)
+          }
+
+          results.push({
+            type: error.code.substr(0, 1) === 'E' ? 'Error' : 'Warning',
```
Comment (Arcanemagus) [inline lib/main.js]: This is missing the possibility of an `Info` type that was originally there.

File: lib/main.js (hunk 2 above)
Comment (Arcanemagus) [inline lib/main.js]: We spawn it as Node, not a system executable, otherwise we would be using the `node_modules/.bin` path :wink:.

**Comment by steelbrain** [comment]
Done

File: lib/main.js [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+          const errorType = error.code.substr(0, 1)
+          const errorLine = error.line - 1
+          let range
+
+          // TODO: Remove workaround of jshint/jshint#2846
+          if (error.character === null) {
+            range = Helpers.rangeFromLineNumber(textEditor, errorLine)
+          } else {
+            let character = error.character
+            // TODO: Remove workaround of jshint/jshint#1457
```
Comment (Arcanemagus) [inline lib/main.js]: This still lists jshint as the source, it's an esprima bug fixed in `esprima@2.7.2`.

Comment (steelbrain) [inline lib/main.js]: I thought I fixed that, sorry about that, fixing again

File: lib/main.js [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+import { CompositeDisposable } from 'atom'
+
+type Linter$Provider = Object
+
+module.exports = {
+  config: {
+    executablePath: {
+      type: 'string',
+      default: Path.join(__dirname, '..', 'node_modules', 'jshint', 'bin', 'jshint'),
+      description: 'Path of the `jshint` executable'
```
Comment (Arcanemagus) [inline lib/main.js]: Again, this needs to be the `jshint` _Node_ script, not a system executable. Maybe something like this?
> 'Path to the `jshint` Node script'

Comment (steelbrain) [inline lib/main.js]: Well, it accepts the path that you get using `which jshint` so I think the title above is pretty fair

**Comment by Arcanemagus** [comment]
LGTM :+1:

**Comment by Arcanemagus** [comment]
Published as v2.0.0. :grinning:

**Comment by steelbrain** [comment]
:dancer:

**Pull request merged**